post: mobile app launch profiling (#124)
getsentry/sentry.engineering: data/blog/mobile-app-launch-profiling.md

---
title: 'Mobile App Launch Profiling'

> date: '2024-04-17'

Author: ```suggestion
date: '2024-09-27'
```

tags: ['mobile', 'ios', 'android', 'profiling']
draft: false
summary: "See what's happening in your app before your first line of code can even run."
images: ['/images/mobile-app-launch-profiling/launch-profile.png']
layout: PostLayout
canonicalUrl: mobile-app-launch-profiling
authors: ['andrewmcknight']
---

> If you use Sentry’s Performance monitoring on iOS/macOS and Android, you have the ability to see [profiles](https://docs.sentry.io/product/profiling/) of how your app runs in the wild on real end user devices. Any time an automatic or manual transaction is recorded, that profiling data is attached. But are there times where you can’t record a transaction? 🤔

Author: ```suggestion
If you use Sentry’s Performance monitoring on Cocoa and Android, you have the ability to see [profiles](https://docs.sentry.io/product/profiling/) of how your app runs in the wild on real end user devices. Any time an automatic or manual transaction is recorded, that profiling data is attached. But are there times where you can’t record a transaction? 🤔
```

Author: @bruno-garcia I'm actually going to revert this and go back to saying Cocoa app. The SDK is named sentry-cocoa, after all.

> If you use Sentry’s Performance monitoring on Cocoa and Android, you also have the ability to see [profiles](https://docs.sentry.io/product/profiling/) of how your app runs in the wild on real end user devices. Any time an automatic or manual transaction is recorded, that profiling data is attached. But are there times where you can’t record a transaction? 🤔

Reviewer: How about we refer to it as `iOS` instead of `cocoa`? Or should it be `iOS/macOS` ?

```suggestion
If you use Sentry’s Performance monitoring on iOS and Android, you have the ability to see [profiles](https://docs.sentry.io/product/profiling/) of how your app runs in the wild on real end user devices. Any time an automatic or manual transaction is recorded, that profiling data is attached. But are there times where you can’t record a transaction? 🤔
```

Author: I'm fine with iOS/macOS!

# App launch performance

<aside>

> 💡 This post will focus on our Cocoa app launch profiling feature. You can read more about the Android counterpart by heading over to that area of our docs: https://docs.sentry.io/platforms/android/profiling/#app-start-profiling

Reviewer: React Native has its own profiler, right? I won't leverage this capability? iOS apps built with Flutter and .NET use the cocoa SDK for profiling as-is:

```suggestion
💡 This post will focus on our iOS app launch profiling feature. Including iOS apps built on [Flutter](https://docs.sentry.io/platforms/flutter/profiling/) and [.NET](https://docs.sentry.io/platforms/dotnet/guides/maui/profiling/). You can read more about the Android counterpart by heading over to [that area of our docs](https://docs.sentry.io/platforms/android/profiling/#app-start-profiling).
```

Author: Hm I'm just not sure, I haven't tested either of those.

Reviewer: They are both on the docs and we're receiving a lot of data (from Flutter in particular, not so much on .NET).

Author: Yeah sounds good.

> <aside>
> 💡 This post will focus on our Cocoa app launch profiling feature. You can read more about the Android counterpart by heading over to that area of our docs: https://docs.sentry.io/platforms/android/profiling/#app-start-profiling
> </aside>
> 
> # App launch performance

Author: ```suggestion
# App launch performance

<aside>
💡 This post will focus on our Cocoa app launch profiling feature. You can read more about the Android counterpart by heading over to that area of our docs: https://docs.sentry.io/platforms/android/profiling/#app-start-profiling
</aside>
```

Apple places special focus on good [performance of an app’s launch](https://developer.apple.com/documentation/xcode/reducing-your-app-s-launch-time). They even built an entire process to enforce good performance: the [watchdog](https://developer.apple.com/documentation/xcode/addressing-watchdog-terminations). There are two ways you can monitor launch performance today:

1. Instruments. You can profile your app while launching it on any devices you have available.
2. MetricKit. You can see a histogram of app launch times in Xcode’s Organizer. This data is inexact, because you can’t correlate any specific context to problematic launches, and the way launch time is measured is out of your control:

> The launch-time metric measures the time from the user tapping the app icon on their Home screen to the app drawing its first frame to the screen. If your app still has to run code after it has drawn its first frame, but before the user can begin using the app, that time doesn’t contribute to the launch-time metric. **Extra startup activities still contribute to the user’s perception of the app’s responsiveness**.

> So, if you or one of your dependencies does some work at launch time, you don’t necessarily see that in the Xcode Organizer.

Reviewer: Is it so that it only measures up to showing the spinner? 
I imagine this is common enough concept that could help folks understand this: 

"Basically this is telling you how quickly you're rendering your loading spinner"

Author: A loading spinner is a specific case, but not everybody necessarily has one. I'm not sure this is necessary to add in.

Reviewer: Even if their app doesn't have a spinner, the concept is very known since apps everyone uses have them. I suggested only to help others undersand (it took me a second to understand what this meant.

Author: So we're actually talking about work being done even before any UI is visible, including something like a loading spinner. I will expand on this a bit.

Author: Yeah you're right, with the more recent pattern being an "empty state" that has some animation.

Author: see https://github.com/getsentry/sentry.engineering/pull/124/files#r1779271955

> So, if you or one of your dependencies does some work at launch time, you don’t necessarily see that in the Xcode Organizer.

Author: ```suggestion
So, if you or one of your dependencies does some work at launch time, you don’t necessarily see that in the Xcode Organizer. These are things like splash screens, loading spinners or empty states with animations indicating that content is on its way.
```

# Sentry app launch profiling

> Now, you can get profiling data from the moment your app starts executing code, before calls to `main` or `appDidFinishLaunching`, where the Sentry SDK is typically initialized and was previously the earliest moment a performance transaction could be recorded.

Reviewer: ```suggestion
Now, once you enable App Launch Profiling, you can get profiling data from the moment your app starts executing code. From before calls to `main` or `appDidFinishLaunching`, which is where the Sentry SDK is typically initialized and was previously the earliest moment a performance transaction could be recorded.
```

Author: This just makes it wordier without adding any additional information. The second sentence also becomes incomplete.

> By enabling `SentryOptions.enableAppLaunchProfiling` in your configuration of `SentrySDK.startWithOptions`, the SDK will automatically profile subsequent launches of your app. If configured, the Sentry profiler is started from one of the SDK’s `+[load]` methods. If you also enabled `SentryOptions.enableTimeToFullDisplayTracing`, the profiling data will be attached to the app start transactions you can find in our App Starts area. Here’s an example of profiling data showing some work being done before `main()` is called, lined up with our app start spans:

Reviewer: ```suggestion
By enabling `SentryOptions.enableAppLaunchProfiling` in your configuration of `SentrySDK.startWithOptions`, the SDK will automatically profile subsequent launches of your app. If configured, the Sentry profiler is started from one of the SDK’s `+[load]` methods. If you also enabled `SentryOptions.enableTimeToFullDisplayTracing`, the profiling data will be attached to the app start transactions you can find in our [App Starts insights](https://docs.sentry.io/product/insights/mobile-vitals/app-starts/). Here’s an example of profiling data showing some work being done before `main()` is called, lined up with our app start spans:
```

> ![A flamechart showing an example launch profile, with work on the main thread before the actual call to `main()`.](/images/mobile-app-launch-profiling/launch-profile.png)

Reviewer: Doesn't seem like our blog expands when clicking on the image. I wonder if we can add that capability, or have a screenshot that's more readable in the current size it renders. (I couldn't read details on it in the preview build, on my curved wide screen monitor)

Author: yeah that's annoying, and the alt text does't appear anywhere. i think we should give some of these things and the stuff i mentioned in the original PR description some love.

i did improve the image though!

In all other cases, a dedicated “launch” transaction will be recorded, where you can access the new profiles.

# Next steps

> Launch profiling can automatically record any launch after you’ve initialized it — but what about the very first launch after a user installs your app? You wouldn’t have had an opportunity to configure launch profiling yet. We considered including a workflow where you’d ship your app with a Plist key to enable launch profiling, so you’d get all the juicy details of what happens on that very first launch. This is still something we’ll consider doing in the future, but for now, we want to battle-test the feature as it is today and make sure it’s a safe and robust part of your app’s launch before adding any additional complexity.

Reviewer: Worth linking to a GitHub ticket about this. So folks reading can engage with this topic if they are keen to have this feature, have opinions or just want to subscribe for updates

Author: Good idea, I'll link the discussion I started to gather feedback. We have no filed issue for this specific idea as we're not sure we'll actually do it. https://github.com/getsentry/sentry-cocoa/discussions/3721

Author: Oh actually I already did that below: https://github.com/getsentry/sentry.engineering/pull/124/files#diff-9e4afedb9e59572b2b80cb82d93552c609951beda62fd565ecd91f961e63749eR46

# What’s happening in your launches?

> Head over to the [docs to enable Launch Profiling](https://docs.sentry.io/platforms/apple/profiling/#enable-launch-profiling) and give app launch profiling a try in your Cocoa and Android apps today — and [let us know what you think](https://github.com/getsentry/sentry-cocoa/discussions/3721)!

Reviewer: ```suggestion
Head over to the [docs to enable Launch Profiling](https://docs.sentry.io/platforms/apple/profiling/#enable-launch-profiling) and give app launch profiling a try in your iOS and Android apps today — and [let us know what you think](https://github.com/getsentry/sentry-cocoa/discussions/3721)!
```

> Head over to the [docs to enable Launch Profiling](https://docs.sentry.io/platforms/apple/profiling/#enable-launch-profiling) and give app launch profiling a try in your iOS and Android apps today — and [let us know what you think](https://github.com/getsentry/sentry-cocoa/discussions/3721)!

Author: ```suggestion
Head over to the [docs to enable Launch Profiling](https://docs.sentry.io/platforms/apple/profiling/#enable-launch-profiling) and give app launch profiling a try in your Cocoa and Android apps today — and [let us know what you think](https://github.com/getsentry/sentry-cocoa/discussions/3721)!
```
